---
title: "ONION"
author: "Monika Piwowar, Wiktor Jurkowski, Ewa Gajda, Cezary Turek"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{ONION}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---



```{r global_options, include=TRUE}
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE)
```

# Table of Contents
i. [About ONION](#about-onion)
i. [ONION workflow](#onion-workflow)
  * 1. [Data input](#input-data)
  * 2. [Cluster molecules with ChEBI Ontology](#chebi-ontology)
  * 3. [Map molecules on Reactome](#use-reactome)
  * 4. [Expand gene clusters with STRING DB](#use-string-db)
  * 5. [Define group](#define-goups)
  * 6. [Lalalala lala](#define-groups-of-genes-and-metabolites)
  * 7. [Multivariate Statistical Analysis](#statistical-analysis)
    * a. [CCA - Canonical Correlation Analysis](#cca---canonical-correlation-analysis)
    * b. [PLS - Partial Least Squares Regression](#pls---partial-least-squares-regression)
ii. [Methods in deatils](#methods-in-deatils)
  * 1. [clasterUsingOntology()](#clasterusingontology())


## About ONION

ONION provides knowedge driven data regularisation to facilitate multivariate analysis of 'omics' data

## ONION Workflow
Below you will find ONION workflow described step by step. By following these steps and using examplary data included in this package user should get the same results as shown in this vignette. 

### Data input
NEW (C) : To start work with ONION you need to provide two 'omics' data sets in data frame form. DF can be created from files as in \example directory.
To start work with ONION you need to provide two 'omics' data sets in tab delimited files with headers (colnames and rownames). You may find examplary data sets in \example directory under package insallation directory (nm-transcriptomics.txt and nm-lipidomics.txt).
Below a small part of nm-lipidomics.txt table is presented just to show users how this file is organized. The first column is called "ChEBI" and contains ChEBI ids for selected small molecules. As for the rest of the columns each of them is labeled with the different name which corresponds with the different samples from which they came from and the numbers in those columns are measured values of molecules mentioned before. Summarazing - the number of rows in this table is equal to the number of tested small molecules (which are identified by their ChEBI ids) and each row contains quantity of that particular molecule (indicated in the first column) measured in different condtions (in different samples). 
Your own input file should be similar and have ChEBI ids or genes symbols in the first column and values of those small molecules/genes measured in different samples in the rest of the columns.
Remember that below is only a small part of nm-lipidomics.txt, original and full file has 21 rows and 41 columns and that is the size of the file that we will use in this vignette. 


```{r, results = 'asis'} 
pathToFileWithChebiIds <- paste(find.package("ONION"),"/example/nm-lipidomics.txt", sep = "")
M <- read.table(pathToFileWithChebiIds, header = TRUE)
pathToExampleFileWithXData <- paste(find.package("ONION"),"/example/nm-transcriptomics.txt", sep = "")
T <- read.table(pathToExampleFileWithXData, header = TRUE)
tab <- head(M,5)
knitr::kable(tab[1:7], caption = "Metabolomics data")
```

### Inititate 

To load ONION run:

```{r, echo=TRUE, eval=TRUE}
library(ONION)
```

```{r, echo=TRUE, results='hide'}
find.package("ONION")
```

### Cluster molecules with ChEBI Ontology

The first step is to load the input file. When this is done then we are ready to map ChEBI identifiers contained in the file to Reactome database by searching of ontologically related molecules present in Reactome's pathways. The output includes six rows of mapped ids.

```{r, echo=TRUE, results='asis'}
clusteredSmallMolecules <- ONION::clusterUsingOntology(chebiIdsDataFrame = M,
                                                           rootColumnName = "ChEBI",
                                                           ontologyRepresentatnion = ONION::firstExistsInReactomeChebiOntology)
    
knitr::kable(head(clusteredSmallMolecules, 6))
```

The result table represent mapping of all small molecules to respective parents and children of ChEBI ontology: "root" denote
s source IDs; 1 - ID already exists in Reactome, if not but there is an alternative in the form of child or parent in ChEBI Ontology, ONION put its id under proper column. If we can not find root id and can not find any parent or children for it, ONION surround it by 0.

Next step is merge of children, roots and parents. For now we have only one method available (ONION::mergeChEBIOntologyWithChildFavoring()), but you can simply develop yourself. Our function merge table with children favoring. 

```{r, echo=FALSE, results='asis'}
mergedSmallMolecules <- ONION::mergeChEBIOntologyWithChildFavoring(clusteredSmallMolecules, rootColumnName = 'root')
knitr::kable(head(mergedSmallMolecules, 6))
#save mapped IDs
merged <- mergedSmallMolecules$ontologyId[mergedSmallMolecules$ontologyId > 0]
write.table(merged, "metabolites_in_Reactome.txt", quote = F, row.names = F, col.names = F)
```

Results include ontologyId column which contains chebi ids used in following procedure: Root column show from which point we start ontology search. In whoWins you can find R, C, P, N letters which mean Root, Child, Parent, None.


### Map molecules on Reactome
Once the result table of merge operation is generated, you can use it to find genes which are in the same pathway as small molecules with specific ChEBI identifiers. Currently the only supported source of canonical pathways is Reactome database.  
Additionally instead of using dataframe from merge result (mergedSmallMolecules) you can use any dataframe with one required column "ontologyId" for example:

```{r, echo=FALSE, results='asis'}
knitr::kable(data.frame(mergedSmallMolecules[1:4, c("ontologyId"), drop = FALSE], XYZ = c("1","2","3","4")))
```

As a result you recive dataframe presented below:

```{r, echo=TRUE, results='hide'}
chebiIdsToReactomePathways <- invisible(ONION::mapReactomePathwaysUnderOrganism(chebiOntologyIds = mergedSmallMolecules[, c("ontologyId"), drop = FALSE], organismTaxonomyId = '9606', idsColumnName = "ontologyId", rootColumnName = NULL))
chebiIdsToReactomePathwaysWithRoot <- invisible(ONION::mapReactomePathwaysUnderOrganism(chebiOntologyIds = mergedSmallMolecules[, c("ontologyId", "root"), drop = FALSE], organismTaxonomyId = '9606', idsColumnName = "ontologyId", rootColumnName = "root"))
```
NEW (C) : Result is DF, but it is complicated (complicated DF, which include lists as cells) object. This DF include:
```{r, echo=FALSE, results='asis'}
head(chebiIdsToReactomePathwaysWithRoot$ontologyId, 1)
# TODO : Convert root to character in code of ONION.
head(as.character(chebiIdsToReactomePathwaysWithRoot$root), 1)
chebiIdsToReactomePathwaysWithRoot$ensembleIds[[1]]
chebiIdsToReactomePathwaysWithRoot$uniProtIds[[1]]
chebiIdsToReactomePathwaysWithRoot$reactomeIds[[1]]
chebiIdsToReactomePathwaysWithRoot$genesSymbolsFromEnsemble[[1]]
chebiIdsToReactomePathwaysWithRoot$genesSymbolsFromUniProt[[1]]
# knitr::kable(head(chebiIdsToReactomePathways, 1))
```


### Use STRING DB

When you have results from Reactome step, then you are ready to use STRING DB step. In this part you search for any extra interactions of gens which you find in Reactome. STRING calls them neighbours. To do it just put results achived from Reactome to `ONION::getStringNeighbours(chebiIdsToReactomePathways)` method. This function produce dataframe. Below one row of this this dataframe is presented:

```{r, echo=FALSE, results='hide'}
chebiIdsToReactomePathwaysAndToStringNeighbours <- invisible(ONION::getStringNeighbours(chebiIdsToReactomePathways[chebiIdsToReactomePathways$ontologyId == "CHEBI:15756",],
                                                                                  stringOrganismId = 9606,
                                                                                  stringDbVersion = "10",
                                                                                  idsColumnName = 'ontologyId',
                                                                                  rootColumnName = NULL,
                                                                                  listOfEnsembleIdColumnName = 'ensembleIds'))
```
NEW (C) : This DF is also a complicated one,
```{r, echo=FALSE, results='asis'}
# knitr::kable(chebiIdsToReactomePathwaysAndToStringNeighbours[chebiIdsToReactomePathwaysAndToStringNeighbours$ontologyId == "CHEBI:15756",])
chebiIdsToReactomePathwaysAndToStringNeighbours$ontologyId
chebiIdsToReactomePathwaysAndToStringNeighbours$ensembleIds[[1]]
chebiIdsToReactomePathwaysAndToStringNeighbours$stringIds[[1]]
chebiIdsToReactomePathwaysAndToStringNeighbours$stringGenesSymbols[[1]]
```

As you can see, this dataframe contains additional columns as stringIds and stringGensSymbols. All of theme are from STRING database and has some edges (connections) with gens from ensembleIds column.

### Define groups of genes and metabolites

This step joins files with data to analyse with results of Reactome step and/or results of STRING step. 
NEW (C) :

### Define goups

Groups are defined in data frame. Dataframe should look like that:
This is similar to GMT file.
```{r, echo=FALSE, results='asis'}
gmtGroupsFilePath <- paste(find.package("ONION"),"/example/nm-groups.txt", sep = "")
groups <- ONION::groupUsingUserDefinition(pathToFileWithGroupDefinition = gmtGroupsFilePath)
knitr::kable(groups)
```


####User selected genes and metabolites
Optionally, the analysis could be executed by hand picked selections of genes and metabolites

```{r, echo=FALSE, results='hide'}
#select small molecules
lip1 <- mergedSmallMolecules[mergedSmallMolecules$root == "CHEBI:27432",]$root
lip2 <- mergedSmallMolecules[mergedSmallMolecules$root == "CHEBI:73705",]$root
joinLip <- c(as.character(lip1), as.character(lip2))

#use Reactome genes mapped to selected small molecules
reactomeTrans1 <- chebiIdsToReactomePathways[chebiIdsToReactomePathways$ontologyId == "CHEBI:15756",]$genesSymbolsFromEnsemble[[1]]
reactomeTrans2 <- chebiIdsToReactomePathways[chebiIdsToReactomePathways$ontologyId == "CHEBI:16015",]$genesSymbolsFromEnsemble[[1]]
joinRecatomeTrans <- c(reactomeTrans1, reactomeTrans2)[!duplicated(c(reactomeTrans1, reactomeTrans2))]
```

### Functional Ineractions DF
You can create a functional interactions data frame by using this method:
```{r, echo=FALSE, results='hide'}
functionalInteractions <- ONION::createFunctionalInteractionsDataFrame(chebiIdsToReactomePathways)
knitr::kable(head(functionalInteractions, 6))
```

### Statistical Analysis

ONION provides two statistical methods to analyse those data:

- CCA - Canonical Correlation Analysis
- PLS - Partial Least Squares Regression

#### CCA - Canonical Correlation Analysis

Calculate CCA by hand:


```{r, echo=FALSE, results='hide'}
    pathToExampleFileWithXData <- paste(find.package("ONION"),"/example/nm-transcriptomics.txt", sep = "")
    pathToExampleFileWithYData <- paste(find.package("ONION"),"/example/nm-lipidomics.txt", sep = "")

    XDF <- read.table(pathToExampleFileWithXData, header = TRUE);
    YDF <- read.table(pathToExampleFileWithYData, header = TRUE);

    ccaResults1 <- ONION::makeCanonicalCorrelationAnalysis(
        xNamesVector = joinRecatomeTrans,
        yNamesVector = joinLip,
            XDataFrame = XDF,
            YDataFrame = YDF)
```


```{r, fig.show='hold', fig.width=6, fig.height=6}
    plotCanonicalCorrelationAnalysisResults(ccaResults = ccaResults1)
```



Calculate CCA on groups:
```{r, echo=FALSE, results='hide'}
    mccReactome <- ONION::makeCCAOnGroups(groupsDefinitionDF = groups, mappingDF = chebiIdsToReactomePathwaysWithRoot, groupsDataDF = YDF, mappingDataDF = XDF)

``` 

Inside the makeCCAOnGroups functions the permutatuin test is automaticly done. To do it yourself on your data do:
```{r, echo=FALSE, results='hide'}
permutationTestsResults <- ONION::makePermutationTestOnCCA(XDataFrame = XDF, YDataFrame = YDF, 17, 2, 50, countedCCA = ccaResults1)
```

To take data from grouped CCA you should know a little the structure, it is like that:
```{r, echo=FALSE, results='hide'}
mccReactome$Molecules[1]
mccReactome$right[[1]]
mccReactome$left[[1]]
mccReactome$ccaResults[[1]]
mccReactome$ccaPermutationTestResults[[1]]
```
```{r, fig.show='hold', fig.width=6, fig.height=6}
plotCanonicalCorrelationAnalysisResults(ccaResults = mccReactome[4,]$ccaResults[[1]])
plotCanonicalCorrelationAnalysisResults(ccaResults = mccReactome$ccaResults[[4]])
```
#### PLS - Partial Least Squares Regression

```{r, echo=FALSE, results='hide'}
PLSResult1 <- ONION::makePartialLeastSquaresRegression(
        joinRecatomeTrans,
        joinLip,
        XDataFrame = XDF,
        YDataFrame = YDF)
```

```{r, fig.show='hold', fig.width=6, fig.height=6}
ONION::plotRmsepForPLS(PLSResult1$training)
```  

```{r, fig.show='hold', fig.width=6, fig.height=6}
ONION::plotRegression(PLSResult1$training, ncompValue = 10)
```

The same but on defined groupd by user.
```{r, echo=FALSE, results='hide'}
groupPlsReactome <- ONION::makePLSOnGroups(groupsDefinitionDF = groups, mappingDF = chebiIdsToReactomePathwaysWithRoot, groupsDataDF = YDF, mappingDataDF = XDF)

```

How to take counted values from results:
```{r, echo=FALSE, results='hide'}
groupPlsReactome$Molecules[1]
groupPlsReactome$right[[1]]
groupPlsReactome$left[[1]]
groupPlsReactome$plsResults[[1]]
groupPlsReactome$plsPermutationTestResults[[1]]
```
How to use it to plot:
```{r, fig.show='hold', fig.width=6, fig.height=6}
    ONION::plotRmsepForPLS(groupPlsReactome$plsResults[[1]]$training)
```
```{r, fig.show='hold', fig.width=6, fig.height=6}
ONION::plotRegression(groupPlsReactome$plsResults[[1]]$training, ncompValue = 10)
```



